### PR TITLE
Address Collision of Contract Address Causes Exceptional Halt

### PR DESCRIPTION
**Pull request by pirapira**
This follows the discussion in [coredev meeting 22](https://github.com/ethereum/pm/issues/20), where no objections were made to the proposal to make contract creation fail in the case of address collision.

### CODE REVIEW

**Comment by cdetrio** [comment]
See also https://github.com/ethereum/EIPs/issues/684

**Comment by pirapira** [comment]
#684 works for me.

**Comment by cdetrio** [comment]
Would be easier to merge this one, can you undelete the branch?

**Comment by pirapira** [comment]
Undeleted.

File: EIPS/address-collision-exceptional-halt.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+## Preamble
+
+    EIP: <to be assigned>
```
Comment (pirapira) [inline EIPS/address-collision-exceptional-halt.md]: Please use 689, and move this file to EIPS/eip-689.md

File: EIPS/address-collision-exceptional-halt.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+It seems unpractical to implement never-used features just for passing tests.  Client implementations will be simpler with this EIP.
+
+## Backwards Compatibility
+
+This EIP is backwards compatible on the main network.
+
+## Test Cases
+
+At least the BlockchainTest called `createJS\_ExampleContract\_d0g0v0\_EIP158` will distinguish clients that implement this EIP.
```
Comment (pirapira) [inline EIPS/address-collision-exceptional-halt.md]: Grammar `implement` should be `implements`.

**Comment by Arachnid** [comment]
This is a courtesy notice to let you know that the format for EIPs has been modified slightly. If you want your draft merged, you will need to make some small changes to how your EIP is formatted:

 - Frontmatter is now contained between lines with only a triple dash ('---')
 - Headers in the frontmatter are now lowercase.

If your PR is editing an existing EIP rather than creating a new one, this has already been done for you, and you need only rebase your PR.

In addition, a continuous build has been setup, which will check your PR against the rules for EIP formatting automatically once you update your PR. This build ensures all required headers are present, as well as performing a number of other checks.

Please rebase your PR against the latest master, and edit your PR to use the above format for frontmatter. For convenience, here's a sample header you can copy and adapt:

```
---
eip: <num>
title: <title>
author: <author>
type: [Standards Track|Informational|Meta]
category: [Core|Networking|Interface|ERC] (for type: Standards Track only)
status: Draft
created: <date>
---
```

File: EIPS/eip-689.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
+## Preamble
+
+---
+eIP: 689
```
Comment (Arachnid) [inline EIPS/eip-689.md]: This needs to be all lowercase.

File: EIPS/eip-689.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
+## Preamble
```
Comment (Arachnid) [inline EIPS/eip-689.md]: These first 3 lines need deleting.

File: EIPS/eip-689.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+Some test cases in the consensus test suite try to deploy a contract at an address already with nonempty code. Although such cases can virtually never happen on the main network before the Constantinople fork block, the test cases detected discrepancies in clients' behavior.  Currently, the Yellow Paper says that the contract creation starts with the empty code and the initial nonce even in the case of address collisions. To simplify the semantics, this EIP proposes that address collisions cause failures of contract creation.
+
+## Motivation
+
+This EIP has no practical relevance to the main net history, but simplifies testing and reasoning.
+
+This EIP has no effects after Constantinople fork because [EIP-86](https://github.com/ethereum/EIPs/pull/208) contains the changes proposed in this EIP. Even before the Constantinople fork, this EIP has no practical relevance because the change is visible only in case of a hash collision of keccak256.
+
+Regarding testing, this EIP releaves clients from supporting reversion of code overwriting.
```
Comment (Arachnid) [inline EIPS/eip-689.md]: s/releaves/relieves/

File: EIPS/eip-689.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+Regarding reasoning, this EIP establishes an invariant that non-empty code is never modified.
+
+## Specification
+
+If `block.number >= 0`, when a contract creation is on an account with non-zero nonce or non-empty code, the creation fails as if init code execution resulted in an exceptional halt.  This applies to contract creation triggered by a contract creation transaction and by CREATE instruction.
+
+## Rationale
+
+It seems unpractical to implement never-used features just for passing tests.  Client implementations will be simpler with this EIP.
```
Comment (Arachnid) [inline EIPS/eip-689.md]: s/unpractical/impractical/

**Comment by pirapira** [comment]
@Arachnid I applied your comments.

**Pull request merged**